File: clrs/algorithms/optimal_bst.py
def optimal_bst(p, q, n):
  e = [[0 for j in range(n + 1)] for i in range(n + 2)]
  w = [[0 for j in range(n + 1)] for i in range(n + 2)]
  root = [[0 for j in range(n + 1)] for i in range(n + 1)]

  # e.g., i = 0, ..., 4; j = 0, ..., 4

  # (1, 0) ... (5, 6)
  for i in range(1, n + 2):
    e[i][i - 1] = q[i - 1]
    w[i][i - 1] = q[i - 1]
  
  for l in range(1, n + 1):
    # (1, 1) ... (5, 5)
    # (1, 2) ... (4, 5)
    # (1, 3) ... (3, 5)
    # (1, 4) ... (2, 5)
    # (1, 5)
    for i in range(1, n - l + 2): 
      j = i + l - 1
      e[i][j] = float('inf')
      w[i][j] = w[i][j - 1] + p[j] + q[j]

      for r in range(i, j + 1):
        t = e[i][r - 1] + e[r + 1][j] + w[i][j]
        
        if t < e[i][j]:
          e[i][j] = t
          root[i][j] = r # r ignores the diagonal
  
  return e, root

def construct_optimal_bst(root, n):
  r = root[1][n]
  print(f"k{r} is the root")

  construct_subtree(root, 1, r - 1, r, 'left')
  construct_subtree(root, r + 1, n, r, 'left')

def construct_subtree(root, i, j, parent, side):
  if i > j:
    print(f"d{j} is the {side} child of k{parent}")
    return

  r = root[i][j]
  print(f"k{r} is the {side} child of k{parent}")

  construct_subtree(root, i, r - 1, r, 'left')
  construct_subtree(root, r + 1, j, r, 'right')

e, root = optimal_bst(
    [0, 0.15, 0.10, 0.05, 0.10, 0.20],
    [0.05, 0.10, 0.05, 0.05, 0.05, 0.10],
    5
)

construct_optimal_bst(root, 5)

File: clrs/algorithms/optimal_bst.js
function optimalBst(successProbabilities, failureProbabilities, n) {
  const expected = Array.from({ length: n + 2 }, () => Array(n + 1).fill(0));
  const weight = Array.from({ length: n + 2 }, () => Array(n + 1).fill(0));
  const root = Array.from({ length: n + 1 }, () => Array(n + 1).fill(0));

  for (let i = 1; i <= n + 1; i += 1) {
    expected[i][i - 1] = failureProbabilities[i - 1];
    weight[i][i - 1] = failureProbabilities[i - 1];
  }

  for (let length = 1; length <= n; length += 1) {
    for (let i = 1; i <= n - length + 1; i += 1) {
      const j = i + length - 1;
      expected[i][j] = Infinity;
      weight[i][j] =
        weight[i][j - 1] + successProbabilities[j] + failureProbabilities[j];

      for (let r = i; r <= j; r += 1) {
        const candidate =
          expected[i][r - 1] + expected[r + 1][j] + weight[i][j];

        if (candidate < expected[i][j]) {
          expected[i][j] = candidate;
          root[i][j] = r;
        }
      }
    }
  }

  return { expected, root };
}

function constructOptimalBst(root, n) {
  const r = root[1][n];
  console.log(`k${r} is the root`);
  constructSubtree(root, 1, r - 1, r, "left");
  constructSubtree(root, r + 1, n, r, "right");
}

function constructSubtree(root, i, j, parent, side) {
  if (i > j) {
    console.log(`d${j} is the ${side} child of k${parent}`);
    return;
  }

  const r = root[i][j];
  console.log(`k${r} is the ${side} child of k${parent}`);
  constructSubtree(root, i, r - 1, r, "left");
  constructSubtree(root, r + 1, j, r, "right");
}

const { root } = optimalBst(
  [0, 0.15, 0.1, 0.05, 0.1, 0.2],
  [0.05, 0.1, 0.05, 0.05, 0.05, 0.1],
  5,
);

constructOptimalBst(root, 5);
